### ISSUE TITLE
Logging with gradient accumulation & DDP

### ISSUE THREAD

**Issue by Tomiinek** [PyTorchLightning/pytorch-lightning#5306]
I have got two questions about logging if using gradient accumulation and DDP:

1. Is there a way to average logged values across the accumulated batches? Logging in training step simply as follows:
    ```python
    def validation_step(self, batch, batch_idx):
        ...
        self.log('val_loss', outputs.loss, sync_dist=True)
        return loss
    ```
    with `accumulate_grad_batches=N` does not work and logs `N` values each 50*`N` steps instead of a single average value each 50 steps.

2. It is recommended to use `sync_dist=True` while using DDP. However, when logging text, images, etc., I would like to use something like:
    ```python
    self.logger.experiment.log({'loss' : ..., 'text' :...}, step=self.global_step)
    ``` 
    but log it only on one device. Is it possible?

Note that I am using the W&B logger.

**Comment by Michiel29** [comment]
I second the first comment. I am using Neptune to log, and for Neptune logging multiple values for the same optimizer step value is not just undesirable but actually breaks the logger.

**Comment by tchaton** [comment]
Hey @Tomiinek, @Michiel29 

For 1. We made this change as we can't assume the logged metrics should just be `averaged` when using accumulated_grad_batches > 1. 
In your example, we actually log every 50 optimizer_step. But the indices will be `50 * N`: num_batches_seen to follow  accumulated_grad_batches=1 logic.

For 2. 

We recommend Logging only on first gpu.

```
from pytorch_lightning.utilities import rank_zero_only


def evaluation_step(...):
    ...

@rank_zero_only
def log_text(self):
     # self.trainer.global_step counts the number time of training_loop.optimizer_step was called.
     self.logger.experiment.log({'text' :...}, step=self.trainer.global_step)
```

Best,
T.C

**Comment by Tomiinek** [comment]
Hi @tchaton , thank you for your response!

**Ad 1** Ok, I see. However, wouldn't be an option to apply a reduction (whatever it is, mean, max, lambda, ...) suitable? The logs for `accumulated_grad_batches=8` currently look like this, i.e. there are 8 steps logged every  `50 * 8` training steps.
![image](https://user-images.githubusercontent.com/13215584/103521988-f7c7ae00-4e79-11eb-852b-5e25272c128f.png)
Do you have a pretty workaround to, for example, average the values that would work when combined with DDP? Something like this but for multiple ranks:
```python
def train_step(self, batch, batch_idx):
    ....
    if self.global_step != old_step:
        average, n = 0, 0
        self.log('val_loss', outputs.loss, sync_dist=True)
    else:
        old_step = self.global_step
        average = (average * n + losss)
        n += 1
        average /= n
```

**Ad 2** Thank you!

**Comment by Tomiinek** [comment]
Moving to #5405
